ast/ast_test.go

```diff
@@ -0,0 +1,247 @@
+			tt.expect(tgt.Name).To(equal("foo"))
+			tt.expect(tgt.Docs).To(equal("foo echoes 'foo'\n"))
+		})
+
+		o.Spec("it parses documentation on later targets", func(tt testCtx) {
+			pers.ConsistentlyReturn(t, tt.reader.NameOutput, "Earthfile")
+			handleMockFile(tt.t, tt.reader, []byte(`
+VERSION 0.6
+
+bar:
+    RUN echo bar
+
+# foo echoes 'foo'
+foo:
+    RUN echo foo
+`))
+			f, err := ast.ParseOpts(context.Background(), ast.FromReader(tt.reader))
+			tt.expect(err).To(not(haveOccurred()))
+
+			tt.expect(f.Targets).To(haveLen(2))
+			tgt := f.Targets[1]
+			tt.expect(tgt.Name).To(equal("foo"))
+			tt.expect(tgt.Docs).To(equal("foo echoes 'foo'\n"))
+		})
+
+		o.Spec("it parses multiline documentation", func(tt testCtx) {
+			pers.ConsistentlyReturn(t, tt.reader.NameOutput, "Earthfile")
+			handleMockFile(tt.t, tt.reader, []byte(`
+VERSION 0.6
+
+# foo echoes 'foo'
+#
+# and that's all.
+foo:
+    RUN echo foo
+`))
+			f, err := ast.ParseOpts(context.Background(), ast.FromReader(tt.reader))
+			tt.expect(err).To(not(haveOccurred()))
+
+			tt.expect(f.Targets).To(haveLen(1))
+			tgt := f.Targets[0]
+			tt.expect(tgt.Name).To(equal("foo"))
+			tt.expect(tgt.Docs).To(equal("foo echoes 'foo'\n\nand that's all.\n"))
+		})
+
+		o.Spec("it does not parse comments with empty lines after them as documentation", func(tt testCtx) {
+			pers.ConsistentlyReturn(t, tt.reader.NameOutput, "Earthfile")
+			handleMockFile(tt.t, tt.reader, []byte(`
+VERSION 0.6
+
+# foo echoes 'foo'
+
+foo:
+    RUN echo foo
+`))
+			f, err := ast.ParseOpts(context.Background(), ast.FromReader(tt.reader))
+			tt.expect(err).To(not(haveOccurred()))
+
+			tt.expect(f.Targets).To(haveLen(1))
+			tgt := f.Targets[0]
+			tt.expect(tgt.Name).To(equal("foo"))
+			tt.expect(tgt.Docs).To(equal(""))
+		})
+
+		o.Spec("it does not check the comment against the target name", func(tt testCtx) {
+			// It felt cleaner to check the doc comment's first word against the
+			// target's name at a higher level where we can display hints to the
+			// user about why the comments are not considered documentation.
+			pers.ConsistentlyReturn(t, tt.reader.NameOutput, "Earthfile")
+			handleMockFile(tt.t, tt.reader, []byte(`
+VERSION 0.6
+
+# echoes 'foo'
+foo:
+    RUN echo foo
+`))
+			f, err := ast.ParseOpts(context.Background(), ast.FromReader(tt.reader))
+			tt.expect(err).To(not(haveOccurred()))
+
+			tt.expect(f.Targets).To(haveLen(1))
+			tgt := f.Targets[0]
+			tt.expect(tgt.Name).To(equal("foo"))
+			tt.expect(tgt.Docs).To(equal("echoes 'foo'\n"))
+		})
+	})
+}
+
+// handleMockFile helps us perform slightly more black-box testing by handling a
+// mockNamedReader as if it were a file-like io.ReadSeeker. This way, we don't
+// need to know in the test how many times the file is seeked back to zero and
+// re-read.
+//
+// This cannot handle non-zero seeks and will fail if it receives a non-zero
+// seek call.
+func handleMockFile(t *testing.T, r *mockNamedReader, body []byte) {
```

Comment:
It took me a bit to understand what

```
		pers.ConsistentlyReturn(t, tt.reader.NameOutput, "Earthfile")
		handleMockFile(tt.t, tt.reader, []byte(`
```
was doing (especially the `ConsistentlyReturn` function as well as where `NameOutput` was coming from -- at first I thought it might have been testing a function called `NameOutput()` consistently returned the value `Earthfile` -- After looking over the `hel` source it makes more sense that it's a mock-helper)

The `handleMockFile` name seems more focused on the implementation  (handling lower-level mocked syscalls) rather than the task that it's doing -- mocking out an Earthfile at a higher-level.

What if we created a new helper function to combine both mock-setup tasks:

```
func mockEarthfile(tt testCtx, earthfileBody []byte) {
    pers.ConsistentlyReturn(tt.t, tt.reader.NameOutput, "Earthfile")
    handleMockFile(tt.t, tt.reader, earthfileBody)
}
```

It could make it more quicker to understand for first time readers (and it'll reduce a line from all the tests).